worker/state/worker.go

```diff
@@ -0,0 +1,219 @@
+					modelUUID,
+					modelStateWorkers,
+					pool,
+				); err != nil {
+					return errors.Trace(err)
+				}
+			}
+		// Useful for tracking down some bugs that occur when
+		// mongo is overloaded.
+		case <-wrenchTimer.C:
+			if wrench.IsActive("state-worker", "io-timeout") {
+				return errors.Errorf("wrench simulating i/o timeout!")
+			}
+			wrenchTimer.Reset(30 * time.Second)
+		}
+	}
+}
+
+// Report conforms to the Dependency Engine Report() interface, giving an opportunity to introspect
+// what is going on at runtime.
+func (w *stateWorker) Report() map[string]interface{} {
+	return w.stTracker.Report()
+}
+
+func (w *stateWorker) processModelLifeChange(
+	modelUUID string,
+	modelStateWorkers map[string]worker.Worker,
+	pool *state.StatePool,
+) error {
+	remove := func() {
+		if w, ok := modelStateWorkers[modelUUID]; ok {
+			w.Kill()
+			delete(modelStateWorkers, modelUUID)
+		}
+		_, _ = pool.Remove(modelUUID)
+	}
+
+	model, hp, err := pool.GetModel(modelUUID)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Model has been removed from state.
+			logger.Debugf("model %q removed from state", modelUUID)
+			remove()
+			return nil
+		}
+		return errors.Trace(err)
+	}
+	defer hp.Release()
+
+	if model.Life() == state.Dead {
+		// Model is Dead, and will soon be removed from state.
+		logger.Debugf("model %q is dead", modelUUID)
+		remove()
+		return nil
+	}
+
+	if modelStateWorkers[modelUUID] == nil {
+		mw := newModelStateWorker(pool, modelUUID, w.pingInterval)
+		modelStateWorkers[modelUUID] = mw
+		_ = w.catacomb.Add(mw)
+	}
+
+	return nil
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *stateWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *stateWorker) Wait() error {
+	err := w.catacomb.Wait()
+	w.cleanupOnce.Do(func() {
+		// Make sure the worker has exited before closing state.
+		if err := w.stTracker.Done(); err != nil {
+			logger.Warningf("error releasing state: %v", err)
+		}
+	})
+	return err
+}
+
+type modelStateWorker struct {
+	tomb         tomb.Tomb
+	pool         *state.StatePool
+	modelUUID    string
+	pingInterval time.Duration
+}
+
+func newModelStateWorker(
+	pool *state.StatePool,
+	modelUUID string,
+	pingInterval time.Duration,
+) worker.Worker {
+	w := &modelStateWorker{
+		pool:         pool,
+		modelUUID:    modelUUID,
+		pingInterval: pingInterval,
+	}
+	w.tomb.Go(w.loop)
+	return w
+}
+
+func (w *modelStateWorker) loop() error {
+	st, err := w.pool.Get(w.modelUUID)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Ignore not found error here, because the pooledState has already
+			// been removed.
+			return nil
+		}
+		return errors.Trace(err)
+	}
+	defer func() {
+		st.Release()
+		_, _ = w.pool.Remove(w.modelUUID)
+	}()
+
+	// Jitter the interval so that each model doesn't attempt to connect to
+	// mongo at potentially the same time.
+	interval := w.pingInterval + jitter(time.Millisecond*200)
```

Comment:
Actually, I can't see how you need the model workers at all...

Reply:
Investigating this further, you need the modelStateWorkers so you can ensure that the pool and the states for the models are correctly released. That does make sense, the problem is that amount of pings can be deafening.

I think the right solution is to offer a channel that can be closed if the root state pool is no longer pingable. That way we only send one ping, rather than N pings.